docs/proposals/horovod-on-tony.md

```diff
@@ -0,0 +1,233 @@
+
+After having a deep understanding of Horovod code and [communicating with developers](https://github.com/horovod/horovod/discussions/2785), i know that the Gloo controller uses a rendezvous server to assign each worker role, and provides HTTP API for workers to obtain cluster information. So each worker  can build a training cluster and start training at the same time.
+
+Horovod is served as two roles, worker and driver. Driver is responsible for starting the rendezvous server and will not participate in training (no GPU required, lightweight). Before starting, driver need to know all workers' hostnames in advance. The worker is only responsible for training. According to TonY's architecture (**Application master** and **task executor**), the design can be as follows.
+
+### Horovod Driver
+__How to start rendezvous server__  
+Reusing Horovod rendezvous server code, we introduce tony-horovod driver launcher to offer a python script
+```python
+# Init the horovod rendezous server
+global_rendezv = RendezvousServer(verbose=1)
+# Output server port, which will be used horovod worker to connect server.
+global_rendezv_port = global_rendezv.start()
+print("Rendezvous server started, port: " + str(global_rendezv_port))
+
+hosts = parse_hosts(worker_list)
+# Output the host plan, it will output local_rank, rank and so on.
+host_alloc_plan = get_host_assignments(hosts, 1)
+
+# Start the server.
+global_rendezv.init(host_alloc_plan)
+```
+
+__When to start driver__  
+After all workers' resource have be assigned and TonY's Application master could get all workers' registry info.
+
+__Where to start driver__  
+Two options  
+1. On TonY application master.  
+This will save resources(no extra resources to start driver), and the amount of code changes will be small. But by injecting relevant Horovod's driver code into AM, it is not elegant.
+2. On TonY task executor.  
+Additional customization of the driver configuration is required and the startup of driver will be covered on TonY automatically. And it is necessary to coordinate the startup sequence between the driver and other workers, because driver should start before worker. 
+
+Second option will be adopted in this PR.  
+In order to unify different machine framework startup, we supposed to create `MLFrameworkRuntime` interface, it will expose methods as follows
+```java
+    /** For AM, getting cluster spec and return to task exectuor **/
+    String constructClusterSpec(String taskId) throws IOException;
+
+    /** For AM, when app finished, it need to call it to release resource **/
+    void destroy();
+
+    /** For AM, init the tony session **/
+    void setTonySession(final TonySession session);
+
+    /** For AM, it ensures that each task executor start sequence. like Horovod driver should start before workers **/
```

Comment:
add a bit more documentation.. it is not clear what can start. Maybe a differnet name for the function as well

Reply:
Got it